cabal.project

```diff
@@ -58,3 +58,7 @@ source-repository-package
   location: https://github.com/input-output-hk/goblins
   tag: cde90a2b27f79187ca8310b6549331e59595e7ba
   --sha256: 17c88rbva3iw82yg9srlxjv2ia5wjb9cyqw44hik565f5v9svnyg
+
+constraints:
```

Comment:
If constraints are needed, they should be specified in the `.cabal` file for the relevant project, rather than in the cabal.project file. `cabal.project` files do not compose.

In this case, I'm not sure why this is needed - `cardano-ledger-alonzo` already compiles quite happily with GHC 8.10...

Reply:
No component depends on ListLike directly, only shelley-spec-ledger-test via process-extras.

This is why it is needed:

```
[20 of 21] Compiling Data.ListLike.Instances ( src/Data/ListLike/Instances.hs, dist/build/Data/ListLike/Instances.o, dist/build/Data/ListLike/Instances.dyn_o )
src/Data/ListLike/Instances.hs:135:10: error:
    Duplicate instance declarations:
      instance IsList BSC.ByteString
        -- Defined at src/Data/ListLike/Instances.hs:135:10
      instance IsList BSC.ByteString
        -- Defined in 'Data.ByteString.Internal'
    |
135 | instance IsList BS.ByteString where
    |          ^^^^^^^^^^^^^^^^^^^^
src/Data/ListLike/Instances.hs:258:10: error:
    Duplicate instance declarations:
      instance IsList BSLC.ByteString
        -- Defined at src/Data/ListLike/Instances.hs:258:10
      instance IsList BSLC.ByteString
        -- Defined in 'Data.ByteString.Lazy.Internal'
    |
258 | instance IsList BSL.ByteString where
    |          ^^^^^^^^^^^^^^^^^^^^^
cabal: Failed to build ListLike-4.7.2 (which is required by lib:test from
cardano-ledger-alonzo-0.1.0.0). See the build log above for details.
```

Well, the problem is not in GHC. It depends on combination of ListLike and bytestring versions. Changed title.

Reply:
We fix an `index-state`, so we should not be seeing newer versions of `bytestring`. How are you trying to build that you are seeing this issue?

Reply:
When I run `cabal build all`, I see
```
Warning: Requested index-state2020-11-15T00:00:00Z is newer than
'hackage.haskell.org'! Falling back to older state (2020-11-14T23:50:07Z).
```

Maybe this is the problem.

Reply:
```
cabal exec -- ghc-pkg field bytestring version
version: 0.10.12.0
```

Reply:
@nc6 It seems that for some reason cabal is choosing an **older** version of `bytesring`.

On `master` I just ran `cabal update` and then `cabal build all` and I see exactly the sate issue as @cblp .

Since `ListLike` is an indirect rather than a direct dependency, adding it to the `cabal.project` file as @cblp suggests is the best solution.

Reply:
Can you verify if this also fixes the problem? https://github.com/input-output-hk/cardano-ledger-specs/pull/2082